inventory_management_system_api/services/system.py

```diff
@@ -93,6 +124,39 @@ def update(self, system_id: str, system: SystemPatchSchema) -> SystemOut:
 
         update_data = system.model_dump(exclude_unset=True)
 
+        parent_id_changing = "parent_id" in update_data and system.parent_id != stored_system.parent_id
+        type_id_changing = "type_id" in update_data and system.type_id != stored_system.type_id
+        if parent_id_changing or type_id_changing:
+            parent_system = None
+            if parent_id_changing:
+                # Parent is being updated
+                if system.parent_id is not None:
+                    parent_system = self._system_repository.get(system.parent_id)
+                    if not parent_system:
+                        raise MissingRecordError(f"No parent system found with ID: {system.parent_id}")
+            elif stored_system.parent_id is not None:
+                # Parent is not being updated but are updating the type so obtain the current parent
+                parent_system = self._system_repository.get(stored_system.parent_id)
+
+            type_id = stored_system.type_id
+            if type_id_changing:
+                # Type is being updated
+                type_id = system.type_id
+
+                if self._system_repository.has_child_elements(system_id):
+                    raise InvalidActionError("Cannot change the type of a system when it has children")
+
+                if not self._system_type_repository.get(system.type_id):
+                    raise MissingRecordError(f"No system type found with ID: {system.type_id}")
+
+            # Ensure the system type matches the parent system type if there is a parent
+            if parent_system is not None and type_id != parent_system.type_id:
+                raise InvalidActionError(
+                    "Cannot move a system into one with a different type"
+                    if parent_id_changing
+                    else "Cannot update the system's type to be different to its parent"
+                )
```

Comment:
Could you please see if some of the logic can be moved into new method(s)?

Reply:
I have moved it to a separate function. I was going to try and create two for it, but I didn't think I could split it very nicely. E.g.
_validate_type_update
_validate_parent_update
Would still need to return the the new/current type or parent in them to do the check at the end combining both so neither would be validating them properly as they are so related.